commons-text/src/main/java/org/jdbi/v3/commonstext/StringSubstitutorTemplateEngine.java

```diff
@@ -35,20 +49,6 @@ public String render(String template, StatementContext ctx) {
         return substitutor.replace(template);
     }
 
-    /**
-     * At the time of writing, defaults means <pre>${foo}</pre> syntax.
-     */
-    public static StringSubstitutorTemplateEngine defaults() {
-        return new StringSubstitutorTemplateEngine(substitutor -> {});
-    }
-
-    /**
-     * Customize the given {@link StringSubstitutor} instance to set your preferred prefix, suffix, escape character, and perhaps other configuration. The instance is created by Jdbi, and is not shared nor re-used. Your customizer function however will be re-used for all instances.
-     */
-    public static StringSubstitutorTemplateEngine withCustomizer(Consumer<StringSubstitutor> customizer) {
-        return new StringSubstitutorTemplateEngine(customizer);
-    }
-
```

Comment:
These methods were released as API in 3.2.0

Reply:
@jdbi/contributors 

I think this highlights that we need a way to do API experiments and include them in releases, without necessarily committing to that API long term.

What if we had a `@Beta` or `@Experimental` annotation that was included in the javadoc? For big features we might put the entire thing in an `experimental` package until it was promoted to a long-term supported feature.

Reply:
I have often wished there was an @Experimental counterpart to @Deprecated... On the other hand, why would this have been marked experimental? It was a done piece of work. Catching things like that is what project conventions and reviews are for...

Tbh I dont think it's a big deal. Everyone understands updates bring changes, and this is still so fresh I doubt a lot of people are using it. It's also not something you need to refer to 500 times in 1 project...

Reply:
> Tbh I dont think it's a big deal. Everyone understands updates bring changes, and this is still so fresh I doubt a lot of people are using it.

That's an assumption, and not as safe as you think. At Eclipse we had a mantra: every API is used by someone. Which means if we make incompatible changes, somebody is going to get mad.

Think about a large software shop. Teams with too many projects on their plate to keep their code groomed regularly. They may only get a chance to update dependencies once or twice a year. When they update from e.g. 3.2.0 to 3.5.0, a semantically versioned project like Jdbi is expected to be fully backward compatible. The only difference users should notice is that bugs are fixed.

> On the other hand, why would this have been marked experimental?

My suggestion is that we start marking _all_ new APIs experimental and give them a few releases to bake. This way, when we start using them in production code we have a grace period for recognizing and fixing design mistakes.

Reply:
> incompatible changes
 
There's a spectrum to that tbh. Replacing a factory method by a constructor or vice versa is a superficial change that doesn't entail functional changes on either side. Ditto for things like a changed package or renamed method, which only takes a search&replace. Completely restructuring an entire mechanism, now that would be annoying.

Philosophy aside, what do I do with these? I understand we want to keep classes as simple as possible and stick to simply exposing a default constructor wherever reasonable. I could mark the outdated factory methods and constants as deprecated for the time being instead of removing them? They are forward compatible...

Reply:
It's a spectrum only if you consider a single app module requiring Jdbi in isolation. It becomes an N-dimensional graph when you factor in third party libraries that integrate with us, e.g. Dropwizard's `dropwizard-jdbi3` artifact.

Take Guava. They have always maintained (and documented) that deprecated APIs will eventually be removed. A year or so ago, they made good on that policy, and removed several deprecated APIs. For some Guava users, this made all hell break loose.

At $JOB, we have some library dependencies that transitively depend on Guava. A few depended on an old version and still used those recently-removed APIs. Others depended on newer API in later versions--added after those other APIs were removed.

This made it impossible to use two such dependencies concurrently in a project, if they depended on incompatible versions of Guava. We had projects stalled for _months_ while we waited for those libraries to update their dependencies to a later version of Guava.

It was an awful mess, and one I will not inflict that on our users.

It is noteworthy that the Guava maintainers, after the outcry over those API removals, vowed never to make binary incompatible changes again.

Please take time to review the Eclipse Foundation's guidelines for API compatibility on their wiki. These were curated from years of trial and error, as well as community feedback:

- https://wiki.eclipse.org/Evolving_Java-based_APIs
- https://wiki.eclipse.org/Evolving_Java-based_APIs_2
- https://wiki.eclipse.org/Evolving_Java-based_APIs_3

Reply:
> Philosophy aside, what do I do with these? I understand we want to keep classes as simple as possible and stick to simply exposing a default constructor wherever reasonable. I could mark the outdated factory methods and constants as deprecated for the time being instead of removing them? They are forward compatible...

Just deprecate, and leave a `@deprecated` javadoc comment directing users to the recommended API alternative.

Reply:
Thanks for putting that so well.  We should strive to avoid breaking API changes unless necessary.  (I still think we can be pragmatic, e.g. a breaking change to test case @Rule just one release after initial merge can slip by occasionally)

I like the idea of @Beta or @Experimental.  I think that lets us lower the bar for inclusion without backing ourselves into a corner.  In particular, it'd let us merge things like the Spring support a little less cautiously.

Reply:
> Just deprecate

Well that's what I've done now.  Now that the build has passed, I think I've addressed all concerns and have still done what I set out to do...

I get your point and I understand the importance of compatibility, but I can't say I think "we will never change anything ever" is completely unitable with maintaining a project. Sometimes things just need to be changed in light of discoveries and problem fixes, or general improvements. You could keep old bug-free classes around and implement breaking improvements in separate clones of those classes (a la `/rest/api/v1`, `/rest/api/v2`, etc), but then you get an awfully polluted project after a while. Jdbi3 broke a lot of stuff coming from v2, so clearly we're not working under a "we will never change anything ever" policy. If someone wants to use 2 libraries in their project, one of which depends on jdbi v2 and the other on v3, they're fucked, just like those guava users were. It happens, it can't be avoided, and while I very much support minimizing breakage, I think a project's coherence takes priority over legacy support. Keep supporting what you can as long as you can, but don't carve problems into stone for the sake of legacy.

But that's just my opinion. This is jdbi and I expect nothing less than for my code to be rejected and changes demanded if something isn't up to jdbi's rules. :)

Anyway, I think we can all be happy with the current changes, no? I'll try to better adjust my code to your conventions right off the bat in the future so we don't need to discuss so much. I have a better view on your opinions and policies now than before, so this thread has definitely been helpful there. :)

Reply:
Changing the packages when we incremented the major version is an intentional decision which ensures that jdbi v2 and v3 can co-exist in the same codebase.  "If someone wants to use 2 libraries in their project, one of which depends on jdbi v2 and the other on v3, they're fucked" -> nope, they both can co-exist totally cleanly!  (as far as I know)

Reply:
Hmm, yeah, I forgot that...

So, can I take away from this then that the rule of thumb is
1) API that comes, stays
2) undesired API must stay supported and may only be marked Deprecated
3) the desired breaking cleanup work can be done at the time of a major version increment, as semver dictates that's when all hell is allowed to break loose

That's basically the net result of all this, right? Maybe we can write this down in the readme for both devs and users to see?